celery/worker/request.py

```diff
@@ -77,100 +77,240 @@ class Request(object):
-        self.argsrepr = headers.get('argsrepr', '')
-        self.kwargsrepr = headers.get('kwargsrepr', '')
-        self.on_ack = on_ack
-        self.on_reject = on_reject
-        self.hostname = hostname or gethostname()
-        self.eventer = eventer
-        self.connection_errors = connection_errors or ()
-        self.task = task or self.app.tasks[type]
-
+        self._argsrepr = self._request_dict.get('argsrepr', '')
+        self._kwargsrepr = self._request_dict.get('kwargsrepr', '')
+        self._on_ack = on_ack
+        self._on_reject = on_reject
+        self._hostname = hostname or gethostname()
+        self._eventer = eventer
+        self._connection_errors = connection_errors or ()
+        self._task = task or self._app.tasks[self._type]
+        
         # timezone means the message is timezone-aware, and the only timezone
         # supported at this point is UTC.
-        eta = headers.get('eta')
+        eta = self._request_dict.get('eta')
         if eta is not None:
             try:
                 eta = maybe_iso8601(eta)
             except (AttributeError, ValueError, TypeError) as exc:
                 raise InvalidTaskError(
                     'invalid ETA value {0!r}: {1}'.format(eta, exc))
-            self.eta = maybe_make_aware(eta, self.tzlocal)
+            self._eta = maybe_make_aware(eta, self.tzlocal)
         else:
-            self.eta = None
-
-        expires = headers.get('expires')
+            self._eta = None
+        
+        expires = self._request_dict.get('expires')
         if expires is not None:
             try:
                 expires = maybe_iso8601(expires)
             except (AttributeError, ValueError, TypeError) as exc:
                 raise InvalidTaskError(
                     'invalid expires value {0!r}: {1}'.format(expires, exc))
-            self.expires = maybe_make_aware(expires, self.tzlocal)
+            self._expires = maybe_make_aware(expires, self.tzlocal)
         else:
-            self.expires = None
-
+            self._expires = None
+        
         delivery_info = message.delivery_info or {}
         properties = message.properties or {}
-        headers.update({
+        self._delivery_info = {
+            'exchange': delivery_info.get('exchange'),
+            'routing_key': delivery_info.get('routing_key'),
+            'priority': properties.get('priority'),
+            'redelivered': delivery_info.get('redelivered'),
+        }
+        self._request_dict.update({
             'reply_to': properties.get('reply_to'),
             'correlation_id': properties.get('correlation_id'),
-            'delivery_info': {
-                'exchange': delivery_info.get('exchange'),
-                'routing_key': delivery_info.get('routing_key'),
-                'priority': properties.get('priority'),
-                'redelivered': delivery_info.get('redelivered'),
-            }
-
+            'hostname': self._hostname,
```

Comment:
Why have you removed this?
I don't think it's a good idea because Celery expects this elsewhere.

Reply:
Actually I don't remove any member of this class, just add two new members "args" and "kwargs" to maintain complete information of args and kwargs sent by Producer.

Reply:
But you removed the `delivery_info` key.

Reply:
self._delivery_info is created at line 155, and is appened to request_dict at line 165. It is exposed as property "delivery_info" at line 173 as well.